File: Algorithms/230. Kth Smallest Element in a BST/javascript/solution.recursion.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} k
 * @return {number}
 */
var kthSmallest = function (root, k) {
    let answer = 0;
    function inorder(root) {
        if (!root) return;
        inorder(root.left);
        if (--k === 0) {
            answer = root.val;
            return;
        }
        inorder(root.right);
    }
    inorder(root);
    return answer;
};

File: Algorithms/230. Kth Smallest Element in a BST/golang/solution.recursion.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func kthSmallest(root *TreeNode, k int) int {
	cur := 0
	val := 0
	dfs(root, k, &cur, &val)
	return val

}

func dfs(node *TreeNode, k int, cur, val *int) {
	if node == nil {
		return
	}
	dfs(node.Left, k, cur, val)

	*cur = *cur + 1
	if *cur == k {
		*val = node.Val
	}
	dfs(node.Right, k, cur, val)
}
